source/common/upstream/default_local_address_selector.h

```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "envoy/upstream/upstream.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Upstream {
+
+/**
+ * Default implementation of UpstreamLocalAddressSelector.
+ */
+class DefaultUpstreamLocalAddressSelector : public UpstreamLocalAddressSelector {
+public:
+  DefaultUpstreamLocalAddressSelector(
+      std::vector<::Envoy::Upstream::UpstreamLocalAddress> upstream_local_addresses);
+
+  // UpstreamLocalAddressSelector
+  UpstreamLocalAddress getUpstreamLocalAddressImpl(
```

Comment:
nit: consider returning a const-ref here if possible.

Reply:
The reason we return by value is that the calling function modifies the socket options on the returned value.